changes/2024-04-08-keystore-discovery/change.md

```diff
@@ -0,0 +1,218 @@
+| Java     | 1.4.0              | n/a             | [KeyStore.java](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/runtimes/java/src/main/smithy-generated/software/amazon/cryptography/keystore/KeyStore.java)     |
+| .NET     | 1.4.0              | n/a             | [KeyStore.cs](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/runtimes/net/Generated/AwsCryptographyKeyStore/KeyStore.cs)                                        |
+
+## Definitions
+
+An "MPL Consumer" is a library, service, or other application
+that uses the AWS Cryptographic Material Providers Library (MPL)
+to manage cryptographic materials.
+An "MPL Consumer" may or may not be using an AWS Crypto Tools product,
+such as the AWS Encryption SDK or AWS Database Encryption SDK.
+
+A "discovery Keystore" is a (Branch) Keystore with
+a KMS Configuration of Discovery.
+
+A "static Keystore" is a (Branch) Keystore with
+a KMS Configuration of KMS Key ARN.
+
+A "strict Keystore" is a (Branch) Keystore with
+a KMS Configuration of MRK KMS Key ARN or KMS Key ARN.
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+Allow MPL Consumers to easily configure
+a Hierarchical Keyring to use
+the KMS Key ARNs already persisted to a Logical Keystore.
+
+Additionally, ensure a KMS Alias is not permitted as a KMS Key ARN.
+
+## Out of Scope
+
+- Multi-region support for a "non-strict" Keystore
+- Administrative/Mutating operations for a Keystore with a KMS Configuration of Discovery;
+  these are `VersionKey` & `CreateKey`.
+- Lifecycle operations for the Keystore
+
+## Motivation
+
+The (Branch) Keystore and the AWS KMS Hierarchical Keyring
+can be used to facilitate Client Side Encryption for
+a Multi-tenant environment where each tenant has a unique
+[Branch Key](../../framework/structures.md#branch-key-materials).
+However, Tenants may want an MPL Consumer to protect
+their data with a particular KMS Key that is NOT necessarily the same
+as other tenants.
+
+Such a Multi-KMS Keystore is, prior to this change, difficult to use
+as a Keystore instance is restricted to one and only one KMS Key ARN.
+Many MPL Consumers have developed a work around to this restriction
+by constructing and then caching a Keystore per KMS Key ARN.
+These MPL Consumers than stand up their own "request-to-Keystore-to-BranchKeyID" logic.
+This workaround may require an additional lookup
+compared to the normal "Encryption Context determines Branch Key ID"
+path.
```

Comment:
Issue: Why did we not consider letting the keyring not just share a CMC?
(on behalf of @seebees )

Reply:
@seebees 

Wait... this a change to the Keystore, there is no Cache involved.

But even if we did change the Cache in this refactor,
that would not have addressed the issue that users need to provide both a Branch Key ID
AND a KMS Key ARN to use a Branch Key.

So... we did not consider letting the keyring share a CMC because that would not have addressed the issue 
at hand.
Branch Keys SHOULD be accessible via Branch Key ID and Keystore Name alone.
The KMS Key ARN is only needed to create or version Branch Keys.

Reply:
Resolved in 83ca2f1